### ISSUE TITLE
Emergency Stop

### ISSUE THREAD

**Issue by Candas1** [Kambodia91/Hoverboard-firmware-hack-FOC-MASTER-SLAVE#1]
### Variant

None

### Control type

None

### Control mode

None

### What can we do to make the firmware better?

Great work.
There is one thought I wanted to share.
Altough the current is limited based on phase current in the FOC model, DC link current is used for current chopping and is faster to react to a fault.
I think the boards you have don't have the DC link current, but an overcurrent trigger.
You could use PB12 as break input.
This would stop PWM output anytime PB12 is triggered.

Example from [vesc](https://github.com/vedderb/bldc/blob/13e60507bfffd2911bb257474ae2fd83f4f51e32/motor/mcpwm_foc.c#L245)

### Describe suggestions or alternatives you have considered

I never tested this but it might be worth investigating.

**Comment by Candas1** [comment]
https://www.st.com/resource/en/application_note/an4277-using-stm32-device-pwm-shutdown-features-for-motor-control-and-digital-power-conversion-stmicroelectronics.pdf

**Comment by Kambodia91** [comment]
Thanks. The board is adapted for emergency shutdown on pin PB12. Disables pwm [output](https://github.com/Kambodia91/Hoverboard-firmware-hack-FOC-MASTER-SLAVE/blob/6fadabe5cf0d9ef7d1e6d02de609948ea8231624/Src/setup.c#L417). In my case I don't know what is the normal operating state on this pin. is it low or high?When I tried to start the engines I turned off this option for testing. When I wanted to turn this option back on, my wheel started to go crazy. Then I did not know that I had a problem with the first launch of the program after uploading it to the processor. Now, every time after uploading the software, I have to run the debug mode, I will manually go through a few lines of code until I get to enable = 1, the program will start working normally.

**Comment by Candas1** [comment]
Unfortunately I don't know, it might depend on the board, so you need to check.

The second point you are mentioning is something different.
Probably it works for you in debug mode because you are introducing a delay as I mentioned in the other thread.
Check if the offset calibration is the problem.

**Comment by Kambodia91** [comment]
I checked the emergency shutdown of the PB12 wheel when it is ok it must be defined as [low](https://github.com/Kambodia91/Hoverboard-firmware-hack-FOC-MASTER-SLAVE/blob/3e544356715d7c48a45d2d0bf42f4ac3d15d25b9/Src/setup.c#L419), when I gave high the wheel did not start.

the second thing with uploading the software, I changed the sampling offset to 4000, so far it's ok. But I can't test it 100% because I have the wheel disconnected from the board. it looks like I upload the software, disconnect from the pc, then connect the wheel holding it in my hand because the wheel is too big and does not fit in the original housing.

it's ok for now, when it happens again, I'll keep looking for the cause, I found another thread, my processor heats up not much but it's warm, on the slave board it's cold. maybe it's damaged

**Comment by LONG-9527** [comment]
我说的你能看懂吗

**Comment by LONG-9527** [comment]
bbcar不能用

**Comment by Candas1** [comment]
> bbcar不能用

If BBcar doesn't work post your issue to bbcar firmware !!!! and not in chinese !!!

**Comment by Kambodia91** [comment]
Program test and control via usart1 with esp32 + blynk application (wifi) https://youtu.be/ViJhMbIGTzU
Foc Control
Speed Mode
Speed max 15 PWM

**Comment by LONG-9527** [comment]
The wheel turns itself and gives off two drops.

**Comment by Kambodia91** [comment]
I control the wheels from my phone for testing.  i use simple blynk app for this (old version) app connects to esp32 via wifi (app <=> cloud blynk <=> esp32 <=> hoverboard)

**Comment by LONG-9527** [comment]
我想要BBCAR

**Comment by Kambodia91** [comment]
see to: https://github.com/larsmm/hoverboard-firmware-hack-FOC-bbcar

**Comment by LONG-9527** [comment]
@Kambodia91 It doesn't work.

**Comment by Kambodia91** [comment]
If BBcar is not working, please report the problem to bbcar [firmware](https://github.com/larsmm/hoverboard-firmware-hack-FOC-bbcar). My version is for separate master/slave boards. Look my picture in [docs](https://github.com/Kambodia91/Hoverboard-firmware-hack-FOC-MASTER-SLAVE/tree/main/docs/pictures) and reed my [RADME.md](https://github.com/Kambodia91/Hoverboard-firmware-hack-FOC-MASTER-SLAVE).

**Comment by Candas1** [comment]
He is not reading anything and wasting everybody's time

**Comment by LONG-9527** [comment]
@Candas1 I came to you for help because I didn't understand、

**Comment by Kambodia91** [comment]
today I uploaded the software and it happened again, the wheel started to go crazy. but it doesn't happen right away, it happens when the processor receives usart communication (esp takes a long time to start). and I can't view the variables (see the offsets) because when this happens the mosfets get very hot. Once I burned one mosfet and a coil in the engine. I didn't have this problem before starting the buzzer, i.e. the buzzer was working but was on another unused output. I'll try to disable it and let you know if it helps.
